typegen/evaluation/file_type_hints_collector.py

```diff
@@ -0,0 +1,159 @@
+        # Track assignments from Functions
+        # NOTE: this handles nested functions too, because the parent reference gets overwritten
+        # NOTE: before we start generating hints for its children
+        self._scope_stack.append(fdef)
+        return True
+
+    def leave_FunctionDef(self, fdef: cst.FunctionDef) -> None:
+        self._scope_stack.pop()
+
+    def visit_Param(self, node: cst.Param) -> bool | None:
+        if not hasattr(node, "annotation"):
+            return True
+        variable_name, line_number = self._get_name_and_line_number(node)
+        type_hint = self._get_annotation_value(node.annotation)
+        self._add_row(line_number, TraceDataCategory.FUNCTION_PARAMETER, variable_name, type_hint)
+        return True
+
+    def visit_FunctionDef_returns(self, node: cst.FunctionDef) -> None:
+        variable_name, line_number = self._get_name_and_line_number(node)
+        if node.returns:
+            type_hint = self._get_annotation_value(node.returns)
+        else:
+            type_hint = None
+        self._add_row(line_number, TraceDataCategory.FUNCTION_RETURN, variable_name, type_hint)
+
+    def visit_AnnAssign(self, node: cst.AnnAssign) -> bool | None:
+        if not hasattr(node, "annotation"):
+            return True
+        _, line_number = self._get_name_and_line_number(node)
+
+        type_hint = self._get_annotation_value(node.annotation)
+        if isinstance(node.target, cst.Attribute):
+            category = TraceDataCategory.CLASS_MEMBER
+            variable_name = node.target.attr.value
+        elif isinstance(node.target, cst.Name):
+            category = TraceDataCategory.LOCAL_VARIABLE
+            variable_name = node.target.value
+        else:
+            raise TypeError("Unhandled case for: " + type(node.annotation.annotation).__name__)
+        self._add_row(line_number, category, variable_name, type_hint)
+        return True
+
+    def _get_name_and_line_number(self, node):
+        pos = self.get_metadata(PositionProvider, node).start
+        variable_name = None if not hasattr(node, "name") else node.name.value
+        variable_line_number = pos.line
+        return variable_name, variable_line_number
+
+    def _add_row(self, line_number: int, category: TraceDataCategory, variable_name: str | None, type_hint: str | None):
+        class_node = self._innermost_class()
+        class_name = None
+        if class_node:
+            class_name = class_node.name.value
+        function_node = self._innermost_class()
+        function_name = None
+        if function_node:
+            function_name = function_node.name.value
+        self.typehint_data.loc[len(self.typehint_data.index)] = [
+            self.file_path,
+            class_name,
+            function_name,
+            line_number,
+            category,
+            variable_name,
+            type_hint,
+        ]
+
+    def _get_annotation_value(self, annotation: cst.Annotation | None) -> str | None:
```

Comment:
This does not cover the case
```
from x import y

def f(inst: y.z): pass
```

Update tests to check this bug (if I wrong, then it is a good test to have at least :) )

Reply:
Added